### DISCUSSION TITLE
Questions about MaHiRo (improved/alternate CFG)

### DISCUSSION THREAD

**Discussion by not-ski** [comfyanonymous/ComfyUI#6066] [Q&A] [answered]
#5975 added "MaHiRo", a post-CFG function that is supposed to essentially improve the way CFG works.

I've tried using this, and generally I have liked the results, but I've noticed something odd: MaHiRo tends to make my CFG behave as if it is "higher" than I've set it.

In other words, if I generate an image at 7.5 CFG _without_ MaHiRo, it is much less saturated than at 7.5 CFG _with_ MaHiRo (seed-controlled, of course).

Is one meant to lower their CFG when using MaHiRo? If so, by how much?

@yoinked-h pinging you for help with this, since you wrote the implementation :)

Thanks in advance!

**Comment by yoinked-h** [comment, marked as answer]
regarding using diff CFG scale, i usually used 4-ish on the model i used (vpred), but use now something like 2.2-2.8 when using MaHiRo
there isnt a concrete number, but as a rule of thumb use like 2 less cfg scale for it 

rundown on MaHiRo:
it, at a high level, takes just following negative guidance, compares it with just following positive guidance, and, if there is low similarity (determined with cossim), merges the old cfg out with the positive guidance [if its completely unrelated (pos == "opposite"  neg), it uses pos, if negative and positive are fully related (pos==neg), it uses cfg and interpolates in between] and returns that as the final guidance
the "paper" i made explains more in detail, but it was moreso a project i did to make cfg less dumb (its smart now, but still bad)


oh, and since people will ask, the backronym is awful and just sucks, moreso since this has no real short name so i had to backronym it terribly
